R/biting_process.R

```diff
@@ -18,30 +18,78 @@ create_biting_process <- function(
   parameters
   ) {
   function(timestep) {
-    # Calculate combined EIR
     age <- get_age(variables$birth$get_values(), timestep)
-
-    bitten_humans <- simulate_bites(
-      renderer,
-      solvers,
-      models,
+    prob_infection <- get_prob_infection(
       variables,
-      events,
-      age,
+      simulate_bites(
+        renderer,
+        solvers,
+        models,
+        variables,
+        events,
+        age,
+        parameters,
+        timestep
+      ),
       parameters,
       timestep
     )
-
+    prob_asymptomatic_progression <- get_prob_progression(
+      'A',
+      variables$state,
+      1 / parameters$da,
+      parameters
+    )
+    prob_subpatent_progression <- get_prob_progression(
+      'U',
+      variables$state,
+      1 / parameters$du,
+      parameters
+    )
+    prob_outcome <- cbind(
+      prob_infection,
+      prob_asymptomatic_progression,
+      prob_subpatent_progression
+    )
+    outcomes <- simulate_competing_outcomes(prob_outcome)
     simulate_infection(
       variables,
       events,
-      bitten_humans,
+      individual::Bitset$new(
+        parameters$human_population
+      )$insert(which(outcomes == 1)),
       age,
       parameters,
       timestep,
       renderer
     )
+    events$subpatent_progression$schedule(which(outcomes == 2), 0)
+    events$recovery$schedule(which(outcomes == 3), 0)
+  }
+}
+
+#' @importFrom stats runif 
+simulate_competing_outcomes <- function(prob_outcome) {
+  prob_occurs <- apply(prob_outcome, 1, function(p) {
+    1 - prod(1 - p)
+  })
```

Comment:
This apply is painfully inefficient, maybe add this little bit to cpp:
```

f1 <- function(prob_outcome){
  prob_occurs <- apply(prob_outcome, 1, function(p) {
    1 - prod(1 - p)
  })
  return(prob_occurs)
}

library(Rcpp)

cppFunction('NumericVector f2(NumericMatrix x) {
  int nrow = x.nrow();
  int ncol = x.ncol();
  NumericVector out(nrow);

  for (int i = 0; i < nrow; i++) {
    double row_product = 1;
    for (int j = 0; j < ncol; j++) {
      row_product *= 1 - x(i,j);
    }
    out[i] = 1 - row_product;
  }
  return out;
}')

test_matrix <- matrix(runif(10000 * 3), ncol = 3)


o1 <- f1(test_matrix)
o2 <- f2(test_matrix)
all.equal(o1, o2)

microbenchmark::microbenchmark(f1(test_matrix), f2(test_matrix))
```

Gives:
```
Unit: microseconds
            expr     min       lq      mean   median       uq     max neval
 f1(test_matrix) 12747.7 14051.25 16822.887 15002.45 17654.85 76961.7   100
 f2(test_matrix)    23.8    26.70    46.031    36.10    40.95   903.9   100
```

Reply:
That's amazing! Tested that and put it in. 5x speedup